### DISCUSSION TITLE
How can I add bots to a local game?

### DISCUSSION THREAD

**Discussion by sanederchik** [efroemling/ballistica#247] [Q&A]
Good evening!

Is it possible to add bots to a local started game?
If yes, how can I do it?
If no, why not to add? It would be a great feature, especially in team games like hockey.

**Comment by JoseANG3L** [comment]
For versions 1.5 there is a mod.
https://youtu.be/LI54zVzp0UE

**Reply by sanederchik** [reply]
> For versions 1.5 there is a mod.
> https://youtu.be/LI54zVzp0UE

Thank you! Unfortunately, I couldn't install mod successfully although I put a `.py` file in a `BombSquad` folder.
I'm afraid it's because of new 1.5 version of BombSquad on Android.

Could you, please, share a link / instruction how to install mods at this version of the game?

**Reply by JoseANG3L** [reply]
First activate the mods folder.

- Go to Settings, Advanced and click on Show mods folder
- Now go to your file manager, find the BombSquad folder and paste the mod.py here
- Open BombSquad, go to Settings, Advanced, Plugins, check the mod box and restart BombSquad
- Ready, you already have the mod

**Reply by sanederchik** [reply]
> First activate the mods folder.
> 
> * Go to Settings, Advanced and click on Show mods folder
> * Now go to your file manager, find the BombSquad folder and paste the mod.py here
> * Open BombSquad, go to Settings, Advanced, Plugins, check the mod box and restart BombSquad
> * Ready, you already have the mod

Thank you for this instruction! Unfortunately, it does not work with my devices. 
I tested both on my `Mi TV 4S 55'` (TV set) and `Xiaomi Redmi Note 9T Pro` (smartphone). I don't see anything in plugins in both cases.
Could it be a platform related issue or is it a global one?

I use a `1.5.29` version of a game in both devices.

**Reply by ghost** [reply]
Can you check a file permissions (rwx)? Maybe game cann't read this plugin?

**Reply by sanederchik** [reply]
> Can you check a file permissions (rwx)? Maybe game cann't read this plugin?

Good evening! I checked it and saw "rw". However, I can't change it without rooting my device which is an extremely risky action. Is there a way to do it without rooting?

**Reply by sanederchik** [reply]
> For versions 1.5 there is a mod.
> https://youtu.be/LI54zVzp0UE

I think I found the reason why it doesn't work with my devices (spoiler: it shouldn't work with any).
The reason is @efroemling introduced a `Plugin` system in `1.5.23` which requires inheriting from a `ba.Plugin` interface and implementing a `on_app_launch` method wrapped inside a `SomePluginName` class.

https://github.com/efroemling/ballistica/wiki/Knowledge-Nuggets#hello-world-2-creating-a-new-plugin

When I made a `myplugin.py` script from his example, it work fined at all of my devices.
I'm afraid all developers should refactor their scripts in order to be workable in current and future releases of this game.

**Comment by TheMikirog** [comment]
Although @JoseANG3L 's is perfectly valid, it doesn't really address the
> It would be a great feature, especially in team games like hockey

type of issue.

I assume you want bots who actually kind of behave like actual players: they know how to play Hockey, know how to avoid danger and obstacles, know how to attack other players and know how to score points.

Why BombSquad doesn't have them yet? Because it was made first and foremost as a local multiplayer experience. If that's not a good enough reason, then let me say that creating these bots is way harder than it might seem. Most modern games nowadays use game engines that already have AI map navigation and bot tools build in to make creating specialized "fake players" much easier, since all the logic responsible for obstacle avoidance, state machines and simple decision making is already there. All you gotta do is utilize them. BombSquad has no such systems and you basically have to create them from scratch.

The bots you see in Coop are the simplest AI you can create without any path finding required. The only thing they know is where they are and where you are, but are totally blind to everything else around them. If you placed these bots into a map like Crag Castle (people already did), you'll notice that they don't know walls exist, they'll blindly walk into mines and bombs and will try to get to you. Bots were designed to be played in Coop, on flat maps and with tons of other bots around them as cannon fodder, so this kind of AI works fairly well in that specific scenario.

If I wanted to make a bot that plays Capture the Flag, I'd need to create / account for:
- A way for bots to move around the map from one place to another effortlessly (create a navigation mesh for each map specifically, Ballistica has no raycasts as far as I'm concerned)
- Make simple jumps to get around the map quicker (if applicable)
- Avoid / defuse land mines
- Be aware of each bomb's lethal range
- Implement melee fighting tactics
- Implement bomb throwing calculations
- Prioritizing targets (who is low on health, who is the biggest threat, should I kill or get the flag)

Just getting the first point on the checklist up and running is a challenge. Even myself who has been modding the game since forever can't really think of a clean and easy way to implement map path finding without some major jank in the code and tons of polishing.
Of course there wasn't a demand for these bots anyway - we have online servers running.
The only reason I'd consider adding this feature is if the servers weren't as populated, but at that point you might be working for nothing.

Needless to say, if implementing such bots was easy, **we would be doing that**.

**Comment by brostosjoined** [comment]
https://github.com/bombsquad-community/plugin-manager/blob/main/plugins/utilities/practice_tools.py thats the closest for now
